Adding CLI support for SQL Transparent Data Encryption (TDE) and TDE …

mihymel opened the pull request:
…with Bring Your Own Key

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

msftclas commented:
@mihymel,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

mihymel commented:
@jaredmoo @nathannfan Please review.

jaredmoo commented:
Server AD Admin test requires a fix in python SDK. This is part of https://github.com/Azure/azure-sdk-for-python/pull/1413

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py:
```diff
@@ -812,11 +818,31 @@ def elastic_pool_list_capabilities(
+        **kwargs):
+
+    if assign_identity is True:
+        kwargs['identity'] = ResourceIdentity(type="SystemAssigned")
```
use enum: `IdentityType.system_assigned.value`

jaredmoo replied:
see https://github.com/Azure/azure-sdk-for-python/blob/c0713004be3670ec30cdfb7e16a2978b95f455c1/azure-mgmt-sql/azure/mgmt/sql/models/sql_management_client_enums.py#L303

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py:
```diff
@@ -812,11 +818,31 @@ def elastic_pool_list_capabilities(
+        assign_identity=False,
+        **kwargs):
+
+    if assign_identity is True:
```
`is True` is redundant

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py:
```diff
@@ -812,11 +818,31 @@ def elastic_pool_list_capabilities(
+        assign_identity=False):
+
+    if instance.identity is None and assign_identity is True:
+        instance.identity = ResourceIdentity(type="SystemAssigned")
```
same comments as above

jaredmoo replied:
Can you comment that identity cannot be removed? Otherwise this code looks suspicious.

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py:
```diff
@@ -910,3 +936,91 @@ def firewall_rule_update(
+        server_name=server_name,
+        key_name=key_name)
+
+def get_server_key_name_from_uri(uri):
```
`_get_server_key_name_from_uri` (leading underscore) to signify that it's a private helper function

jaredmoo replied:
Would be nice if there was a quick comment to explain the purpose of this function e.g. "The SQL server key API requires that the server key has a specific name based on the vault, key name, and key version".

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py:
```diff
@@ -910,3 +936,91 @@ def firewall_rule_update(
+
+def get_server_key_name_from_uri(uri):
+
+    match = re.match('https(.)+\.vault(.)+\/keys\/[^\/]+\/[0-9a-zA-Z]+$', uri)
```
should there be a `^` at the start of this regex to match the start of the string? so that you don't match a string with leading junk like `"blah blha hello world https://YourVaultName.vault.azure.net/keys/YourKeyName/01234567890123456789012345678901"`

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py:
```diff
@@ -910,3 +936,91 @@ def firewall_rule_update(
+        client,
+        resource_group_name,
+        server_name,
+        server_key_type,
```
So --server-key-type must be specified even if you only want to modify the key id, correct? This is not necessarily bad, just want to make sure this was a conscious choice.

mihymel replied:
Correct. Even if the server is already using a Key Vault key as encryption protector, we still want them to provide server key type argument if they are simply changing to a different Key Vault key.

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/params.py:
```diff
@@ -453,6 +455,19 @@ def _configure_security_policy_storage_params(cmd):
+#####
+
+with ParametersContext(command='sql db tde') as c:
+    c.argument('database_name', options_list=('--database', '-d'), required=True,
```
nit: since this is multi line function call, please put each arg on a separate line so that it looks pretty

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/params.py:
```diff
@@ -579,6 +594,9 @@ def _configure_security_policy_storage_params(cmd):
     })
 
+    c.argument('assign_identity', options_list=('--assign-identity', '-i'),
+                help='Assign an identity for the server.')
```
Can you explain this a little more? Imagine I have a general idea of what AAD is, but I don't know how server identity relates to AAD.

mihymel replied:
I can update this to the same help message we use in powershell. What do you think of:
"Generate and assign an Azure Active Directory Identity for this server for use with key management services like Azure KeyVault."

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/params.py:
```diff
@@ -618,9 +637,7 @@ def _configure_security_policy_storage_params(cmd):
-    c.argument('server_name',
-               options_list=('--server', '-s'),
-               help='The name of the Azure SQL server.')
+    c.argument('server_name', arg_type=server_param_type)
```
👍

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/params.py:
```diff
@@ -635,3 +652,28 @@ def _configure_security_policy_storage_params(cmd):
+    c.argument('key_name', options_list=('--name', '-n'))
+    c.argument('key_id', options_list=('--key-id', '-k'),
+               required=True,
+               help='The Azure Key Vault keyid of the server key')
```
👍

jaredmoo replied:
Is "keyid" standard terminology that is also used elsewhere?

mihymel replied:
We use "keyid" in the corresponding powershell cmdlet. Azure Key Vault uses "kid" or "Key Identifier". I'm considering switching this to "kid" to be consistent with AKV, but I'm concerned that "kid" will be confusing outside the context of Key Vault. I'll discuss with my team.

mihymel replied:
Discussed with my team... it's best to be consistent with AKV here. I'll modify this to be "kid".

jaredmoo commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/test_sql_commands.py:
```diff
@@ -76,27 +78,39 @@ def test_sql_server_mgmt(self, resource_group, resource_group_location):
+                     JMESPathCheck('name', server1),
+                     JMESPathCheck('resourceGroup', rg),
+                     JMESPathCheck('administratorLogin', user),
+                     JMESPathCheck('identity.type', 'SystemAssigned')])
```
👍

jaredmoo commented:
Minor comments, overall looks great. Can you paste output for `-h` that shows all affected commands? e.g. `az sql server -h`, `az sql server key -h`, etc ?

jaredmoo commented:
Run this and fix all issues in order to unbreak the build: `.\scripts\check_style.bat --module sql --pep8 --pylint`

If you are using visual studio code, enable this preference to auto fix trailing whitespace warningsthese when saving files: 
"files.trimTrailingWhitespace": true

mihymel commented:
@jaredmoo  I had to make some changes to the Transparent Data Encryption code as well as the Encryption Protector code after pulling in the latest sdk. Can you take a second pass of this PR?

 In the SDK, these were both "singletons" whose operations used to live under their parent resources operations class, but now have their own class.

tjprescott commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py:
```diff
@@ -78,17 +82,28 @@
+    with ServiceGroup(__name__, get_sql_database_transparent_data_encryptions_operations,
+                      transparent_data_encryptions_operations, custom_path) as s:
+        with s.group('sql db tde') as c:
+            c.command('update', 'create_or_update')
```
Update has a very specific meaning in CLI 2.0.  Since this is likely a "create or replace" style PUT, this command should be called "set".

jaredmoo replied:
Ok, makes sense but I didn't get this feedback when I added `sql db threat-policy update` or `sql db audit-policy update`. Those both have optional updateable properties thought, so I still think `update` using `c.generic_update_command` makes sense for them. Thoughts?

tjprescott replied:
I'm fine with that.  The key thing is that update implies "I update what I want and everything else stays the same as it was."  If I use "set" that implies that whatever I don't specify is null.  So if the command preserves the existing values (which generic update definitely does), then update is definitely the right choice.

tjprescott commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py:
```diff
@@ -78,17 +82,28 @@
+    with ServiceGroup(__name__, get_sql_database_transparent_data_encryption_activities_operations,
+                      transparent_data_encryption_activities_operations, custom_path) as s:
+        with s.group('sql db tde-activity') as c:
+            c.command('list', 'list_by_configuration')
```
With there be additional tde-activity commands?  If not, this should be up-leveled `sql db list-tde-activity`.

mihymel replied:
There will not be additional tde-activity commands. I'll change it to your suggestion.

jaredmoo replied:
Or `sql db tde list-activity` ?

jaredmoo replied:
Since there is `sql db tde` group already

tjprescott replied:
@jaredmoo yes, that makes even more sense!

tjprescott commented on src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py:
```diff
@@ -216,6 +231,27 @@
+        with s.group('sql server encryption-protector') as c:
+            c.command('show', 'get')
+            c.command('list', 'list_by_server')
+            c.custom_command('update', 'encryption_protector_update')
```
I didn't see the implementation of this, but since there is not a `create` command, it leads me to believe this should be called `set`.  The distinction:

`create`: traditional ARM PUT.  Effectively a create or replace.
`update`: uses an ARM PATCH or simulates PATCH behind the scenes. If it has a "create or patch" behavior, then this is the appropriate verb too.
`set`: Basically a PUT that is used in the absence of separate create and update commands.

jaredmoo replied:
@mihymel is it possible that there might be some updateable properties and therefore might want to have an `update` command in the future? If yes, to me it makes more sense to have an `update` command now rather than `set` now and also `update` later. But you might disagree, just my 2c

mihymel replied:
You can think of the encryption protector just as "the server key that will be used for transparent data encryption". With this in mind, you can understand that encryption protector probably won't need updateable properties in the future. I think `set` makes a lot of sense.

jaredmoo replied:
Ok. That sounds fine. Are `show` and `list_by_server` both needed? What happens if you `show` on a fresh server that you have never `set` before?

jaredmoo replied:
Just brainstorming here: what do you think about commands like `sql db tde set-key` , `sql db tde show-key` ?

mihymel commented:
Updated the commands based on feedback. The updated commands are below:

Server Key related:
az sql server key create
az sql server key show
az sql server key list
az sql server key delete
 
Encryption Protector  (renamed to tde-key) related:
Az sql server tde-key set 
Az sql server tde-key show
 
Transparent Data Encryption related:
Az sql db tde set
Az sql db tde show
Az sql db tde list-activity

tjprescott commented:
I'm happy with the command names. @jaredmoo?

jaredmoo commented:
Commands look good to me. Thx

tjprescott commented:
Please rebase and we can merge this :)

mihymel commented:
@tjprescott Apologies for the delay. I rebased, updated the commands to match my previous message, and rerecorded the affected tests. I also condensed the commits into 1 commit.

tjprescott commented on src/command_modules/azure-cli-sql/HISTORY.rst:
```diff
@@ -6,6 +6,7 @@ Release History
 Unreleased
 ++++++++++++++++++++
 * az sql server list --resource-group argument is now optional. If not specified, all sql servers in the entire subscription will be returned.
+* Adding support for SQL Transparent Data Encryption (TDE) and TDE with Bring Your Own Key
```
The other two changes listed here were in fact released as 2.0.12. Please fix this bad merge and then we'll get this in.
